opm/output/data/Wells.hpp

```diff
@@ -145,9 +162,35 @@ namespace Opm {
             return completion->rates.get( m, 0.0 );
         }
 
+        template <class MessageBufferType>
```

Comment:
Where did this `MessageBufferType` come from?

Reply:
It is set in collecttoiorank.hh in ewoms/ebos

        typedef Dune :: Point2PointCommunicator< Dune :: SimpleMessageBuffer > P2PCommunicatorType;
        typedef typename P2PCommunicatorType :: MessageBufferType MessageBufferType;

I wanted to add the communication for the data::well in the same class to make it easier to remember to add the communication part when adding new data to data::wells. I have added a test in ebos (test_ecl_output.cc) to make sure everything is communicated.

Reply:
I **really really** dislike this way of tricking the dependency graph using templates. ewoms depends on opm-ouptut. now you introduce symbols in opm-output which makes opm-output depend on ewoms in a circular fashion, and the only reason this works is that the code in opm-output is dead and not accessed before we reach ewoms. I will not merge this in the present form.

My main gripe with the current code is that you introduced a hidden dependency - and now have methods in the well object *which can not be tested*; I certainly see the rationale for doing it like this - but then we should in my opinion just suck it up - and make opm-output depend explicitly on the dune communicator type - because that is now the case. A minimum solution would be to move the `test_ecl_output.cc` from ebos to opm-output and then either make the test implementation depend explicitly on the dune commonicator, or use mock messagebuffer - I guess `std::stringstream` would do.

Reply:
To make opm-output depend on the Dune communicator should be avoided. I realize now that this changes the thinking that opm-output is parallel-unaware, and we should not do that lightly. The minimum solution suggested by @joakim-hove is ok (with a trivial messagebuffer used for testing, not depending on the ewoms class or moving it), but I think there may be better solutions. The simplest alternative is to put the read() and write() methods as free functions in ebos, although that would not resolve the problem you try to solve here (that the class gets more members and the buffer read/write function updates are forgotten).

Reply:
Ok, I see your points. With the current solution in data::well I really don't like to move the read() and write() methods to ebos. A more generic data::well container, more like data::solution, would make moving the read() and write() methods to ebos less problematic in my opinion, then the communication code could be generic as well without need for frequent updates. This would certainly require a lot of work both in the opm-output code and in ebos. If the minimum solution is acceptable for both of you, I would like to go for that. @joakim-hove I guess you don't want me to move the whole test_ecl_output.cc to opm-output, but the test that explicitly tests the read() and write() methods. For that I think extending the test_summary is sufficient.

Reply:
I have an idea that could work, but I have not thought it 100% through:

To separate "accessing specific data members" from parallel communication, the opm-output classes could define operator<< and operator>>, (alternatively call them read() and write()) which would be completely normal I/O functions, writing to standard streams. I would suggest ASCII is acceptable for the limited amount of data related to wells, but we could make it binary. Then the communication code just uses stringstreams to get in/out these data (that part does not care about the contents other than it being X bytes).

To build the global object, we would need the read() method to append, not overwrite the existing content. So perhaps we should call it append() instead.

Bonus benefit: easier print debugging. (Not a big one...)

Reply:
ASCII is essentially not acceptable here. And then your suggestions with stringstream becomes quite cumbersome and error prone. @totto82 suggestion is quite simple and easy to understand. The template argument with MessageBuffer can be easily tested locally.

Reply:
> The template argument with MessageBuffer can be easily tested locally.

With a trivial MessageBuffer class I assume? That is an acceptable solution for me as I wrote earlier, I just tried to see if there were alternative solutions.

Reply:
I have added read/write test in test_summary. Is this what you had in mind?

Reply:
> I have added read/write test in test_summary. Is this what you had in mind?

Yes - this was what I had in mind.